simplified-cardcreator/src/main/java/org/nypl/simplified/cardcreator/viewmodel/ConfirmationViewModel.kt

```diff
@@ -0,0 +1,69 @@
+package org.nypl.simplified.cardcreator.viewmodel
+
+import android.graphics.Bitmap
+import android.net.Uri
+import android.os.Environment
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+import org.nypl.simplified.cardcreator.viewmodel.ConfirmationEvent.CardConfirmed
+import org.nypl.simplified.cardcreator.viewmodel.ConfirmationEvent.SaveCardPermissionsCheck
+import org.slf4j.LoggerFactory
+import java.io.File
+import java.io.FileOutputStream
+
+class ConfirmationViewModel(
+  initialState: ConfirmationData,
+) : ViewModel() {
+  private val logger = LoggerFactory.getLogger(ConfirmationViewModel::class.java)
+
+  private val _events = MutableSharedFlow<ConfirmationEvent>()
+  val events = _events.asSharedFlow()
```

Comment:
Yeah you are right that can lose events right now. I guess I didn't think it was a big deal given there's no logic being executed - they're just UI events routed through the view model rather than events representing a side effect of work the view model is doing. So worst case the user just hits the button again. Still not a great experience though and probably worth addressing. Also maybe a case for handling those directly in the view rather than through the viewmodel. 

I could probably just add some replay to the SharedFlow and fix most of the issue, though that still introduces some other potential problems. 

I did some reading and it seems like as of coroutines 1.4.0 Channels don't fully prevent dropped events either. There's an interesting thread discussing single events that mentions it [here](https://gist.github.com/gmk57/330a7d214f5d710811c6b5ce27ceedaa?permalink_comment_id=3744320#gistcomment-3744320). For the code in question we are sometimes emitting events in a background thread, so I think the conversation issue is pertinent. 

Recently Google put forth their own [recommended approach](https://developer.android.com/topic/architecture/ui-layer/events#handle-viewmodel-events) for single events. They advocate exposing events as part of the viewmodel state, rather than a separate stream, and to have the view notify the viewmodel when events have been processed. I'd like to try their recommendation even though I'm sure it'll change again in a year.

[Here's](https://proandroiddev.com/sending-view-model-events-to-the-ui-eef76bdd632c) a good article talking about Google's new recommended approach.

Reply:
Interesting reading about Channels ! It looks like there is no issue if events are sent on the main thread though, which is not a big deal. Your new approach looks totally fine too. It's more complex, but more obvious to prove to be correct as well.